File: framework/languages/java/samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/handlers/EnsureCustomerActorHandler.java
package systems.zlink.samples.deliverydispatch.server.customergateway.handlers;

import systems.zlink.framework.actors.ActorRefSnapshot;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.channels.ZLinkRequestContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;
import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup("customer-route")
public final class EnsureCustomerActorHandler
    implements ZLinkRequestHandler<Messages.EnsureCustomerActorReq, Messages.EnsureCustomerActorRes> {
    private final ZLinkActorManager actors;

    public EnsureCustomerActorHandler(ZLinkActorManager actors) {
        this.actors = actors;
    }

    @Override
    public CompletionStage<Messages.EnsureCustomerActorRes> handle(
        Messages.EnsureCustomerActorReq request,
        ZLinkRequestContext context) {
        return actors.getOrCreate(request.customerId(), SampleNames.CustomerActorType, request)
            .thenApply(actor -> new Messages.EnsureCustomerActorRes(
                request.customerId(), ActorRefSnapshot.from(actor)));
    }
}

File: framework/languages/java/samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/handlers/EnsureCustomerActorHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway.handlers

import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.framework.actors.ActorRef
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.channels.ZLinkRequestContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.actors.ActorRefSnapshot
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.EnsureCustomerActorRes
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.EnsureCustomerActorReq

@ZLinkHandlerGroup("customer-route")
class EnsureCustomerActorHandler(
    private val actors: ZLinkActorManager,
) : ZLinkSuspendingRequestHandler<EnsureCustomerActorReq, EnsureCustomerActorRes> {
    override suspend fun handle(
        request: EnsureCustomerActorReq,
        context: ZLinkRequestContext,
    ): EnsureCustomerActorRes {
        val actor = actors.getOrCreate(request.customerId, SampleNames.CustomerActorType, request).await()
        return EnsureCustomerActorRes(
            customerId = request.customerId,
            actorRef = ActorRefSnapshot.from(actor),
        )
    }
}
